### PR TITLE
Minor: Assert `streaming_merge` has non empty sort exprs

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/7794

## Rationale for this change

While debugging https://github.com/apache/arrow-datafusion/issues/7794 (which I think is a bad plan) the lack of Sort exprs caused very strange behavior that made it harder to debug.

Specifically, the merge code ended up passing bad indices to `interleave` which resulted in a panic far away from the source of the problem. 

## What changes are included in this PR?

1. Return an internal error if no sort exprs are passed to `streaming_merge`
2. Assert in `RowCursor` that there is at least one row
3. Add some docs

## Are these changes tested?
Yes by existing tests

## Are there any user-facing changes?
No, only easier to debug failures

### CODE REVIEW

**Comment by alamb** [comment]
I verified this assert fires for my test case

**Comment by alamb** [comment]
@tustvold  and @wiedld might you have a few minutes to review this PR?

File: datafusion/physical-plan/src/sorts/cursor.rs [hunk 1]
```diff
@@ -48,16 +48,17 @@ impl std::fmt::Debug for RowCursor {
 
 impl RowCursor {
     /// Create a new SortKeyCursor from `rows` and a `reservation`
-    /// that tracks its memory.
+    /// that tracks its memory. Thre must be at least one row
```
Comment (jackwener) [inline datafusion/physical-plan/src/sorts/cursor.rs:51]:
```suggestion
    /// that tracks its memory. There must be at least one row
```

File: datafusion/physical-plan/src/sorts/cursor.rs [hunk 2]
```diff
@@ -48,16 +48,17 @@ impl std::fmt::Debug for RowCursor {
...
     ///
     /// Panic's if the reservation is not for exactly `rows.size()`
```
Comment (jackwener) [inline datafusion/physical-plan/src/sorts/cursor.rs:53]:
```suggestion
    /// Panics if the reservation is not for exactly `rows.size()`
```

**Review by jackwener — APPROVED** [review]
Thanks @alamb

File: datafusion/physical-plan/src/sorts/cursor.rs [hunk 3]
```diff
@@ -207,8 +211,12 @@ pub struct FieldCursor<T: FieldValues> {
 }
 
 impl<T: FieldValues> FieldCursor<T> {
-    /// Create a new [`FieldCursor`] from the provided `values` sorted according to `options`
+    /// Create a new [`FieldCursor`] from the provided `values` sorted according
+    /// to `options`.
+    ///
+    /// Panics if the array is empty
     pub fn new<A: FieldArray<Values = T>>(options: SortOptions, array: &A) -> Self {
+        assert!(array.len() > 0, "Empty array passed to FieldValues");
```
Comment (viirya) [inline datafusion/physical-plan/src/sorts/cursor.rs:219]:
```suggestion
        assert!(array.len() > 0, "Empty array passed to FieldCursor");
```

File: datafusion/physical-plan/src/sorts/streaming_merge.rs [hunk 4]
```diff
@@ -60,6 +60,11 @@ pub fn streaming_merge(
     fetch: Option<usize>,
     reservation: MemoryReservation,
 ) -> Result<SendableRecordBatchStream> {
+    // If there are no sort expressions, preserving the order
+    // doesn't mean anything (and result in inifinite loops)
```
Comment (viirya) [inline datafusion/physical-plan/src/sorts/streaming_merge.rs:64]:
```suggestion
    // doesn't mean anything (and result in infinite loops)
```

**Review by viirya — APPROVED** [review]

**Pull request merged**